src/rgw/rgw_tools.cc

```diff
@@ -101,6 +101,23 @@ int rgw_init_ioctx(const DoutPrefixProvider *dpp,
   return 0;
 }
 
+uint32_t rgw_bucket_shard_index(const rgw_obj_key& obj_key,
+				int num_shards)
+{
+  std::string sharding_key;
+  if (obj_key.ns == RGW_OBJ_NS_MULTIPART) {
+    RGWMPObj mp;
+    mp.from_meta(obj_key.name);
+    sharding_key = mp.get_key();
+  } else {
+    sharding_key = obj_key.name;
+  }
+
+  uint32_t sid = ceph_str_hash_linux(sharding_key.c_str(), sharding_key.size());
+  uint32_t sid2 = sid ^ ((sid & 0xFF) << 24);
+  return rgw_shards_mod(sid2, num_shards);
```

Comment:
can use `RGWSI_BucketIndex_RADOS::bucket_shard_index()` to avoid duplicating this stuff

and the dependency on `RGWSI_BucketIndex_RADOS` helps to make it clear that this function is rados-specific